Fix race condition

fabioxgn opened the pull request:
@toolateforteddy there was a race condition writing to the response handler. As this function is implemented by various protocols and we can't be sure if they can handle multiple messages and how fast they are, I've made 2 changes:

- Added a mutex to write one message at a time, so the messages will be queued and sent synchronously to the protocols
- Changed the buffer length to 100 as the messages are mostly like text and I don't thing that having a 100 message on the buffer will cause any issues

toolateforteddy commented on bot.go:
```diff
@@ -127,11 +130,17 @@ func (b *Bot) SendMessage(target string, message string, sender *User) {
 		select {
 		case msg := <-b.msgsToSend:
-			b.handlers.Response(msg.target, msg.message, msg.sender)
+			b.sendResponse(msg.target, msg.message, msg.sender)
```
This isn't a race. Only one goroutine will ever be calling messageSender, so there is nothing to synchronize between.

This is an unnecessary lock and unlock.

fabioxgn replied:
The passive commands run in a goroutine and can call the SendMessage at the same time. If you run `go test --race` it will point that out.

But I think that you are correct in this case, that the Response method, at least in production will not be called at the same time.

But in the tests, as they are called synchronously it is creating a race condition.

I'll think about some way to fix the tests as having the tests break becausa of a race condition that will not be in production isn't a good thing.

toolateforteddy commented on bot.go:
```diff
@@ -127,11 +130,17 @@ func (b *Bot) SendMessage(target string, message string, sender *User) {
+func (b *Bot) sendResponse(target, message string, sender *User) {
+	b.mutex.Lock()
+	b.handlers.Response(target, message, sender)
+	b.mutex.Unlock()
```
This code should be: ```
b.mutex.Lock()
defer b.mutex.Unlock()
b.handlers.Response(target, message, sender)
```
This way it is exception safe.

That said, this mutex is unnecessary. The only thing that ever calls sendResponse will be a single goroutine.

fabioxgn replied:
That's right, I'll fix that.

fabioxgn commented:
@toolateforteddy did some refactoring. Now the race condition is fixed and I removed the `synchronousMessageSending` option because I think that we should test the production code and not have a different behaviour for the tests.

In the tests now I use a channel to receive the message and implemented a method to wait for the expected amount of messages.

I'm not happy about 2 tests which I had to put a `time.Sleep`, but this tests are testing if the bot doesn't reply to anything, so right now I don't have a better idea how to test this.

toolateforteddy commented on cmd_test.go:
```diff
@@ -22,29 +26,51 @@ const (
+		case <-time.After(1 * time.Second):
+			t.Fatal("Timed the shit out")
+		}
+		if len(msgs) == count && len(errs) == errorCount {
```
Can these equal signs be made to be >= ? And or add a check for them being too large, and fail the test inside this method if either one grows beyond it's expected length.

toolateforteddy commented on cmd_test.go:
```diff
@@ -22,29 +26,51 @@ const (
+		case err := <-cmdError:
+			errs = append(errs, err)
+		case <-time.After(1 * time.Second):
+			t.Fatal("Timed the shit out")
```
Can this print out msgs and errors so far received?

fabioxgn replied:
@toolateforteddy yep, I think that this is a good idea, also forgot to clean up this message. Too much 🍺

toolateforteddy commented on cmd_test.go:
```diff
@@ -173,87 +195,87 @@ func TestDisabledCommands(t *testing.T) {
-	b.synchronousMessageSending = true
 	b.MessageReceived(&ChannelData{Channel: "#go-bot"}, &Message{Text: "!cmd"}, &User{Nick: "user"})
 
+	waitMessages(t, 1, 1)
```
I really appreciate that after this method has completed, you know that there is at least this many elements in the specified channels.

toolateforteddy commented on telegram/telegram.go:
```diff
@@ -21,7 +21,6 @@ func responseHandler(target string, message string, sender *bot.User) {
 		return
 	}
 	msg := tgbotapi.NewMessage(id, message)
-	msg.ReplyToMessageID = msg.ReplyToMessageID
```
Why did you kill this?

fabioxgn replied:
The linter hinted that it was assigned to itself, I don't see the point of this code.

toolateforteddy replied:
Oh totally. Kill it with fire.
